### DISCUSSION TITLE
how to do a many-to-many relationship?

### DISCUSSION THREAD

**Discussion by dxgi** [prisma/prisma#21500] [Q&A] [answered]
![image](https://github.com/prisma/prisma/assets/61899298/637ffbae-96fa-4258-84e6-bc92a354a74d)

Explanation: Both Post-1 and Post-2 are using Game-1 and Game-2.

How would I make any posts be able to use any games even though if a post is already using those games. The following schema did not seem to work (many-to-many??)

```prisma
// schema.prisma
model User {
    ...
    posts Post[]
    games Game[]
}

model Game {
    ...
    User        User?    @relation(fields: [userId], references: [id], onDelete: Cascade)
    userId      String?
    Post        Post[]
}

model Post {
    ...
    games Game[]
    User        User?    @relation(fields: [userId], references: [id], onDelete: Cascade)
    userId      String?
}
```

**Comment by dxgi** [comment]
I could use string of game ID's instead but as much as possible I want to run findFirst with include (games: true)

```
model Post {
    games     String[]
    User        User?    @relation(fields: [userId], references: [id], onDelete: Cascade)
    userId     String?
}
```

**Comment by ludralph** [comment, marked as answer]
Hi @dxgi 👋 

Thank you for raising this question.

You can define an [implicit many-to-many](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/many-to-many-relations#implicit-many-to-many-relations) relationship by simply listing the related model as an array on both sides. Here's how you can adjust your schema:
```prisma
model User {
  id    Int    @id @default(autoincrement())
  posts Post[]
  games Game[]
}

model Game {
  id     Int    @id @default(autoincrement())
  posts  Post[]
  User   User?  @relation(fields: [userId], references: [id])
  userId Int?
}

model Post {
  id     Int    @id @default(autoincrement())
  games  Game[]
  User   User?  @relation(fields: [userId], references: [id])
  userId Int?
}
```
A `Game` has a list of `Post` and `Post` has a list of `Game`, which Prisma interprets as a many-to-many relationship. Prisma will handle the creation of the relation table internally.

If you need to store additional fields on the relation, such as metadata about the relationship, you would then create an explicit many-to-many relationship by defining a separate model to act as the relation table. 
```prisma
model User {
  id    Int    @id @default(autoincrement())
  posts Post[]
  games Game[]
}

model Post {
  id     Int          @id @default(autoincrement())
  games  GameOnPost[]
  user   User?        @relation(fields: [userId], references: [id])
  userId Int?
}

model Game {
  id     Int          @id @default(autoincrement())
  User   User?        @relation(fields: [userId], references: [id])
  userId Int?
  posts  GameOnPost[]
}

model GameOnPost {
  post   Post @relation(fields: [postId], references: [id])
  postId Int
  game   Game @relation(fields: [gameId], references: [id])
  gameId Int

  @@id([postId, gameId])
}
```

With this schema, you can use the `findFirst` method with `include` to fetch related games:
```js
const postWithGames = await prisma.post.findFirst({
  where: {
    // ... your conditions ...
  },
  include: {
    games: true,
  },
});
```

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏
